main.go

```diff
@@ -418,23 +157,53 @@ func main() {
 	// NB: We SHOULD fetch the table columns ONLY after the call to EnsureCompatibleExportTable.
 	// The EnsureCompatibleExportTable function potentially alters the schema of the export table in the client warehouse.
 	tableColumns := wh.GetExportTableColumns()
-	for {
-		lastSyncedRecord, err := wh.LastSyncPoint()
-		if BackoffOnError(err) {
-			continue
-		}
 
-		numBundles, err := ProcessExportsSince(wh, tableColumns, lastSyncedRecord, exportProcessor)
-		if BackoffOnError(err) {
-			continue
-		}
+	recordTransform := func(rec map[string]interface{}) ([]string, error) {
+		return TransformExportJSONRecord(rec, tableColumns, wh.ValueToString)
+	}
 
-		// if we processed any bundles, there may be more - check until nothing comes back
-		if numBundles > 0 {
-			continue
+	// Outer loop handles restarting after issues in the pipeline. Some of these may be recoverable, so we want to just
+	// try to restart the pipeline from the last checkpoint.
+	for {
+		// Get the last saved checkpoint
+		startTime, err := wh.LastSyncPoint()
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		log.Printf("Looking for bundles since %s", startTime.Format(time.RFC3339))
+
+		p := pipeline.NewPipeline(conf, recordTransform)
+		savedExports, errs := p.Start(startTime)
+
+		for {
+			select {
+			case savedExport := <-savedExports:
+				log.Printf("Bundle saved to: %s", savedExport.Filename)
+
+				// Try to upload the bundle to the warehouse and retry with back off if it fails
+				for {
+					err := LoadBundles(wh, savedExport.Filename, savedExport.Meta...)
```

Comment:
It does assume that `fs.ExportList()` returns exports in the correct order (oldest to newest). We specify that that is the case in the [list export API endpoint](https://help.fullstory.com/develop-rest/data-export-api).

The pipeline handles things serially after that. Since the channels are unbuffered, they will move in lock step, preventing any ordering issues.